### ISSUE TITLE
Fails to handle ssl.conf on (some?) Ubuntu vivid systems

### ISSUE THREAD

**Issue by pde** [letsencrypt/letsencrypt#981]
`sudo venv/bin/letsencrypt -t --debug -vvvvv`
...
2015-10-14 19:36:29,059:DEBUG:letsencrypt.plugins.disco:Other error:(PluginEntryPoint#apache): ('There has been an error in parsing the file (%s): %s', u'/etc/apache2/mods-enabled/ssl.conf', u'Syntax error')
...
I'm not sure ssl.conf has been modified on this system.  Anyway:
`cat /etc/apache2/mods-enabled/ssl.conf`

```
IfModule mod_ssl.c>

        # Pseudo Random Number Generator (PRNG):
        # Configure one or more sources to seed the PRNG of the SSL library.
        # The seed data should be of good random quality.
        # WARNING! On some platforms /dev/random blocks if not enough entropy
        # is available. This means you then cannot use the /dev/random device
        # because it would lead to very long connection times (as long as
        # it requires to make more entropy available). But usually those
        # platforms additionally provide a /dev/urandom device which doesn't
        # block. So, if available, use this one instead. Read the mod_ssl User
        # Manual for more details.
        #
        SSLRandomSeed startup builtin
        SSLRandomSeed startup file:/dev/urandom 512
        SSLRandomSeed connect builtin
        SSLRandomSeed connect file:/dev/urandom 512

        ##
        ##  SSL Global Context
        ##
        ##  All SSL configuration in this context applies both to
        ##  the main server and all SSL-enabled virtual hosts.
        ##

        #
        #   Some MIME-types for downloading Certificates and CRLs
        #
        AddType application/x-x509-ca-cert .crt
        AddType application/x-pkcs7-crl .crl

        #   Pass Phrase Dialog:
        #   Configure the pass phrase gathering process.
        #   The filtering dialog program (`builtin' is a internal
        #   terminal dialog) has to provide the pass phrase on stdout.
        SSLPassPhraseDialog  exec:/usr/share/apache2/ask-for-passphrase

        #   Inter-Process Session Cache:
        #   Configure the SSL Session Cache: First the mechanism 
        #   to use and second the expiring timeout (in seconds).
        #   (The mechanism dbm has known memory leaks and should not be used).
        #SSLSessionCache                 dbm:${APACHE_RUN_DIR}/ssl_scache
        SSLSessionCache         shmcb:${APACHE_RUN_DIR}/ssl_scache(512000)
        SSLSessionCacheTimeout  300

        #   Semaphore:
        #   Configure the path to the mutual exclusion semaphore the
        #   SSL engine uses internally for inter-process synchronization. 
        #   (Disabled by default, the global Mutex directive consolidates by default
        #   this)
        #Mutex file:${APACHE_LOCK_DIR}/ssl_mutex ssl-cache


        #   SSL Cipher Suite:
        #   List the ciphers that the client is permitted to negotiate. See the
        #   ciphers(1) man page from the openssl package for list of all available
        #   options.
        #   Enable only secure ciphers:
        SSLCipherSuite HIGH:!aNULL

        # SSL server cipher order preference:
        # Use server priorities for cipher algorithm choice.
        # Clients may prefer lower grade encryption.  You should enable this
        # option if you want to enforce stronger encryption, and can afford
        # the CPU cost, and did not override SSLCipherSuite in a way that puts
        # insecure ciphers first.
        # Default: Off
        #SSLHonorCipherOrder on

        #   The protocols to enable.
        #   Available values: all, SSLv3, TLSv1, TLSv1.1, TLSv1.2
        #   SSL v2  is no longer supported
        SSLProtocol all -SSLv3

        #   Allow insecure renegotiation with clients which do not yet support the
        #   secure renegotiation protocol. Default: Off
        #SSLInsecureRenegotiation on

        #   Whether to forbid non-SNI clients to access name based virtual hosts.
        #   Default: Off
        #SSLStrictSNIVHostCheck On

</IfModule>

# vim: syntax=apache ts=4 sw=4 sts=4 sr noet
```

**Comment by jdkasten** [comment]
@pde  You are missing a a ```<``` before IfModule.

**Comment by jdkasten** [comment]
^ Was this a copy and paste error?

**Comment by jdkasten** [comment]
I haven't tested this out... but if Apache is that relaxed in its handling of configuration files.  (Which at this point wouldn't completely surprise me)  I think we can probably make the bug report more helpful, by trying to detect what went wrong, but I don't think we should try to autocorrect these errors at this time.

**Comment by pde** [comment]
Sorry, that is a copy and paste error.  Re-pasting from this system:

```
<IfModule mod_ssl.c>

        # Pseudo Random Number Generator (PRNG):
        # Configure one or more sources to seed the PRNG of the SSL library.
        # The seed data should be of good random quality.
        # WARNING! On some platforms /dev/random blocks if not enough entropy
        # is available. This means you then cannot use the /dev/random device
        # because it would lead to very long connection times (as long as
        # it requires to make more entropy available). But usually those
        # platforms additionally provide a /dev/urandom device which doesn't
        # block. So, if available, use this one instead. Read the mod_ssl User
        # Manual for more details.
        #
        SSLRandomSeed startup builtin
        SSLRandomSeed startup file:/dev/urandom 512
        SSLRandomSeed connect builtin
        SSLRandomSeed connect file:/dev/urandom 512

        ##
        ##  SSL Global Context
        ##
        ##  All SSL configuration in this context applies both to
        ##  the main server and all SSL-enabled virtual hosts.
        ##

        #
        #   Some MIME-types for downloading Certificates and CRLs
        #
        AddType application/x-x509-ca-cert .crt
        AddType application/x-pkcs7-crl .crl

        #   Pass Phrase Dialog:
        #   Configure the pass phrase gathering process.
        #   The filtering dialog program (`builtin' is a internal
        #   terminal dialog) has to provide the pass phrase on stdout.
        SSLPassPhraseDialog  exec:/usr/share/apache2/ask-for-passphrase

        #   Inter-Process Session Cache:
        #   Configure the SSL Session Cache: First the mechanism 
        #   to use and second the expiring timeout (in seconds).
        #   (The mechanism dbm has known memory leaks and should not be used).
        #SSLSessionCache                 dbm:${APACHE_RUN_DIR}/ssl_scache
        SSLSessionCache         shmcb:${APACHE_RUN_DIR}/ssl_scache(512000)
        SSLSessionCacheTimeout  300

        #   Semaphore:
        #   Configure the path to the mutual exclusion semaphore the
        #   SSL engine uses internally for inter-process synchronization. 
        #   (Disabled by default, the global Mutex directive consolidates by default
        #   this)
        #Mutex file:${APACHE_LOCK_DIR}/ssl_mutex ssl-cache


        #   SSL Cipher Suite:
        #   List the ciphers that the client is permitted to negotiate. See the
        #   ciphers(1) man page from the openssl package for list of all available
        #   options.
        #   Enable only secure ciphers:
        SSLCipherSuite HIGH:!aNULL

        # SSL server cipher order preference:
        # Use server priorities for cipher algorithm choice.
        # Clients may prefer lower grade encryption.  You should enable this
        # option if you want to enforce stronger encryption, and can afford
        # the CPU cost, and did not override SSLCipherSuite in a way that puts
        # insecure ciphers first.
        # Default: Off
        #SSLHonorCipherOrder on

        #   The protocols to enable.
        #   Available values: all, SSLv3, TLSv1, TLSv1.1, TLSv1.2
        #   SSL v2  is no longer supported
        SSLProtocol all -SSLv3

        #   Allow insecure renegotiation with clients which do not yet support the
        #   secure renegotiation protocol. Default: Off
        #SSLInsecureRenegotiation on

        #   Whether to forbid non-SNI clients to access name based virtual hosts.
        #   Default: Off
        #SSLStrictSNIVHostCheck On

</IfModule>

# vim: syntax=apache ts=4 sw=4 sts=4 sr noet
```

**Comment by jdkasten** [comment]
Related issue: https://github.com/hercules-team/augeas/issues/191

This is a problem with the Httpd lens in Augeas 1.3 -- (I wasn't seeing the problem as I was personally running 1.2).  The problem has been resolved in Augeas 1.4

The problem arises when comments are left at the end.  For letsencrypt-apache, we might want to disallow Augeas 1.3 as this will certainly be a frequent problem.

The other alternative is not to hard-fail on parsing issues.  The client would have still successfully run on the other files in this case and it wouldn't have affected the end result.  This would complicate tracking down other bugs later.

**Comment by pde** [comment]
There are likely to be a lot of OSes out there that have this bug, unpatched.  Fedora seems to have gone to 1.4 bu Ubuntu [doesn't have a ticket for this problem](https://bugs.launchpad.net/ubuntu/+source/augeas/+bugs?field.status:list=NEW).

I presume there's no way for us to work around the issue by feeding in files with the trailing line sripped off?

**Comment by pde** [comment]
Or shipping the lense from 1.4 ourselves?  They're just text files, right?

**Comment by pde** [comment]
Apparently we can [pass lenses in](https://github.com/hercules-team/python-augeas/blob/master/augeas.py#L103).

**Comment by erebus2** [comment]
+1
(sorry have to do this because of https://github.com/isaacs/github/issues/283)

**Comment by erebus2** [comment]
I think that shipping our own lens would just be a work-around that should be avoided if possible. IMO, the right thing to do would be to depend on augeas-lenses v1.4. This would force distros to fix their augeas packages before they can ship letsencrypt.

See also: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=802665

**Comment by jdkasten** [comment]
@schwartzcr Would you like to take a look at this?

Here is where augeas is initialized right now.  We don't automatically load any modules.
https://github.com/letsencrypt/letsencrypt/blob/master/letsencrypt-apache/letsencrypt_apache/augeas_configurator.py#L32

Here is where the Httpd lens is actually loaded. (The lens is added on the first parsing of any Httpd files)
https://github.com/letsencrypt/letsencrypt/blob/master/letsencrypt-apache/letsencrypt_apache/parser.py#L493

**Comment by pde** [comment]
@erebus2 agree it's a workaround, but our primary aim for the time being is to get people working copies of the letsencrypt client, not to persuade distributions to fix old buggy packages.  IME the latter can be a painful and thankless task...

**Comment by SwartzCr** [comment]
@erebus2 can you elaborate on why you think it'd be a bad idea to specify our own lense?
As far as the status of the package, it looks like the bug was patched in April https://github.com/hercules-team/augeas/pull/220
Debian never upgraded to 1.3 https://packages.debian.org/search?searchon=names&keywords=augeas-lenses
Ubuntu has 2 distros that are 1.3: http://packages.ubuntu.com/search?keywords=augeas-lenses
and fedora is on 1.4: http://koji.fedoraproject.org/koji/packageinfo?packageID=6131
@pde did you run into this problem on Ubuntu?
So one solution is to bother ubuntu to fix this. @hlieberman says that the package for LE should be fine as he'll make it depend on 1.2 which ubuntu will copy, but I feel that this leaves open the possible future issue of someone trying to build from github and manually getting the wrong libaugaes0 and breaking it.
Altreantely if the Debian and Ubuntu packages for LE were packaged correctly then it'd work out of the box for most people.
As for @jdkasten's suggestion that we disallow 1.3, that's not really possible. We can specify a version in out _deb_common.sh but if that isn't the version that the package manager is serving then it just won't install. We could altrenately try to pin to a specific other package repository, but that might be more trouble than just loading the lense.

All that said I'm in favor of opening a bug in Ubuntu's (https://bugs.launchpad.net/ubuntu/+source/augeas/+bug/1510318) and then making sure we specify a version for our package(s)

**Comment by domcleal** [comment]
I've uploaded patches based on both Augeas 1.2.0 and 1.3.0 to the Debian and Ubuntu bug trackers respectively.  The Ubuntu issue looks easy to get involved with - if you can take the patch, put it into a patch against the source package and work with Robie to get the fix tested and pushed out.  Hopefully one of the Debian maintainers will pick up the patch too.

If I can be of any more help from an Augeas dev or Fedora perspective, please let me know.  It's great to see LE use Augeas, so I hope we can get it fixed.

**Comment by SwartzCr** [comment]
Looks like it's been fixed for Ubuntu: https://bugs.launchpad.net/ubuntu/+source/augeas/+bug/1510318

**Comment by pde** [comment]
@SwartzCr I encountered it on an Ubuntu 15.04 system with libaugeas0 1.3.0-0ubuntu1.

I think that making our pip-distributed copy of the client immune to this problem is the best option; having it explicitly conflict with broken augeas versions via the bootstrap scripts is a slightly inferior but workable answer.

**Comment by bmw** [comment]
Just spun up a Digital Ocean droplet with Ubuntu 15.04 and can confirm this is still a problem :(

**Comment by domcleal** [comment]
It hasn't been fixed on Ubuntu 15.04, it's marked as "triaged" only - the patch is only current applied to Ubuntu 16.04 (Xenial).  See Robie's [comments 6 and 8](https://bugs.launchpad.net/ubuntu/+source/augeas/+bug/1510318/comments/6) about testing the Xenial package, [available from here](http://packages.ubuntu.com/xenial/augeas-lenses).

If somebody confirms on the Ubuntu ticket that it's fixing the problem, then it can be backported to the stable releases.

**Comment by SwartzCr** [comment]
@domcleal so all we need to do is test this on 16.04 and report back?
@bmw do you want to take on that task?

**Comment by basak** [comment]
@SwartzCr if can you test and verify on 16.04 then I'll prepare updates to 15.04 and 15.10 for you. Once the proposed updates to 15.04 and 15.10 have been reviewed they will again individually need testing by someone before they land.

I hope that sounds reasonable. Our processes result from a need to reduce risk of regression to existing users.

**Comment by erebus2** [comment]
@SwartzCr Well, I just don't like it because it's a workaround and not an actual fix.

However, due to #1202, we might have to ship our own lens anyway.

**Comment by SwartzCr** [comment]
do we have a lens we want to use or should I just load the httpd lens from Augaes 1.2?

**Comment by erebus2** [comment]
No, the lens from 1.2 is broken as well. On my machine, it incorrectly detects a syntax error in /etc/nagios/apache2.conf

**Comment by SwartzCr** [comment]
do we have a working lens?

**Comment by pde** [comment]
If we don't, then the best option is to take then lens from 1.4 and start fixing bugs in it :(

**Comment by SwartzCr** [comment]
does anyone have a link to the httpd.lns file from v1.4? All I can find in the tarball is httpd.aug, and for various reasons I can't find a good place to install it to try to rip the lens file from /augeas/load/Httpd/lens

**Comment by pde** [comment]
I believe the .aug files might actually be the lenses?  At least those are the files contained in the `augeas-lenses` debian package...

**Comment by domcleal** [comment]
The .aug file is the lens, the .lns extension is kind of an internal reference.  The file httpd.aug provides the lens Httpd.lns.

If you want to load it in preference to the one in your distribution package, install it to /usr/share/augeas/lenses/, or put it in a directory and then set the environment variable AUGEAS_LENS_LIB.  The dir can also be passed as the second argument to the Augeas constructor (`augeas.Augeas(loadpath="/example")`).

**Comment by SwartzCr** [comment]
@domcleal yup, that's what I've done: https://github.com/SwartzCr/letsencrypt/blob/c0048d7480e7ba8b62baed3bfded465146c85156/letsencrypt-apache/letsencrypt_apache/augeas_configurator.py#L36

**Comment by SwartzCr** [comment]
@domcleal do you have a sense if python augeas will support loadpaths for augeas 1.0? We're finding that Travis is failing for this change because the travis machine runs ubunut 12 which has augeas 1.0

**Comment by domcleal** [comment]
@SwartzCr should be fine, it's had the functionality for years

**Comment by pde** [comment]
Closed by #1291! Great work @SwartzCr, @kuba and @ArchimedesPi !
